### ISSUE TITLE
Proactive messaging using Contextual Bandit

### ISSUE THREAD

**Issue by marinam34** [NikPeg/telegram-gpt-template#18]
## Summary

Сейчас бот может отправлять проактивные сообщения пользователю по фиксированному расписанию. Нужно реализовать механизм, при котором бот будет решать, стоит ли писать пользователю сейчас, и какой тип сообщения выбрать, на основе контекста пользователя и его прошлых реакций.

Идея: использовать Contextual Multi-Armed Bandit как модуль выбора стратегии проактивного сообщения. LLM при этом не должна сама решать всю логику проактивности. Bandit выбирает стратегию, а LLM только формулирует конкретный текст сообщения.

## Why

Фиксированные сообщения по расписанию могут быстро надоедать пользователю. Нам нужно, чтобы бот постепенно адаптировался под каждого пользователя:

* как часто ему писать;
* в какое время ему лучше писать;
* какой стиль сообщений ему подходит;
* когда лучше не писать вообще;
* какие типы сообщений вызывают хороший диалог;
* какие сообщения пользователь игнорирует или воспринимает негативно.

## Reference

В похожей health-related работе авторы использовали гибридный подход cMAB + LLM для персонализированных motivational messages в приложении для физической активности:

https://arxiv.org/html/2506.07275v4

Что там сделали:

* contextual multi-armed bandit выбирал тип intervention-сообщения;
* LLM генерировала конкретный текст внутри выбранного типа;
* reward считался по тому, насколько пользователь счёл сообщение полезным/релевантным;

Хотим адаптировать эту идею под нашего AI-psychologist bot: bandit выбирает тип проактивного сообщения и принимает решение, писать ли его, а LLM генерирует текст с учётом памяти пользователя.

## What Contextual Bandit should decide

Например, Bandit должен выбрать одно из следующих действий:

* `do_nothing` — не писать сейчас;
* `short_checkin` — короткий вопрос “как ты?”;
* `soft_support` — мягкое поддерживающее сообщение;
* `follow_up_previous_topic` — вернуться к прошлой теме пользователя;
* `calendar_followup` — спросить про событие из календаря;
* `health_stress_support` — поддержка при сигнале стресса/плохого сна, если health integration будет доступна;
* `practical_advice` — предложить маленькое практическое действие;
* `emotional_reflection` — помочь пользователю осмыслить состояние.

Важно: `do_nothing`. Модель должна иметь возможность не писать вообще ничего.

## Context features

Context Builder должен собирать объект с признаками для каждого пользователя в каждой точке принятия решения. Можно подумать какие признаки для этого лучше использовать, хорошо бы иметь исследование на эту тему, но для m4p можно какие то базовые взять.

Пример:

{
  "days_since_last_user_message": 3,
  "hour": 20,
  "is_evening": true,
  "messages_sent_today": 0,
  "ignored_last_proactive": false,
 "ignored_count": 20,
  "ignored_count_last_week": 1,
  "avg_reply_delay_minutes": 35,
  "last_mood_score": 4,
  "recent_negative_topic": true,
  "has_upcoming_calendar_event": true,
  "has_finished_calendar_event": false,
  "health_stress_signal": null,
  "preferred_short_messages": true
}

## Reward calculation

После каждого проактивного сообщения нужно рассчитывать награду на основе реакции пользователя.

например, награды для MVP могут так рассчитываться:

+1, если пользователь ответил;
+1, если пользователь ответил быстро;
+2, если после сообщения диалог продолжился на несколько сообщений;
+2, если пользователь явно отметил сообщение как полезное;
-1, если пользователь проигнорировал сообщение;
-2, если пользователь отреагировал негативно;
-5, если пользователь отключил проактивные сообщения.

Но нельзя оптимизировать модель только  под факт ответа пользователя. Модель не должна учиться отправлять манипулятивные или усиливающие тревогу сообщения только потому, что пользователи на них чаще реагируют.

## MVP implementation

Для первой версии не нужна сложная ML-модель.

MVP может использовать:

веса действий для каждого пользователя;
стратегию decaying epsilon-greedy или более умное сэмплирование Томпсона;
простое обновление награды.

Пример весов действий:

{
  "short_checkin": 0.5,
  "soft_support": 0.5,
  "follow_up_previous_topic": 0.5,
  "calendar_followup": 0.5,
  "practical_advice": 0.5,
  "do_nothing": 0.5
}

Правило обновления:

new_weight = old_weight + learning_rate * reward

Альтернативный более плавный вариант обновления:

Q_new = Q_old + alpha * (reward - Q_old)

Предлагаемые стартовые значения:

alpha = 0.1;
epsilon = 0.2 для новых пользователей;
постепенно снижать epsilon до 0.05–0.1 после достаточного количества взаимодействий.

## Decision points

Scheduler может запускать policy несколько раз в день, например:

утром;
днём;
вечером;
после события в календаре;
перед важным событием в календаре;
после сигнала о стрессе из health-данных, если такая интеграция доступна.

В каждой точке принятия решения бандит выбирает:

* ничего не делать;
* или отправить один из разрешённых типов проактивного сообщения.

## Safety constraints

* не отправлять сообщения ночью;
* не отправлять больше 1 проактивного сообщения в день, если пользователь не ответил;
* не отправлять сообщения после 2–3 проигнорированных проактивных сообщений подряд;
* не использовать тревожные или манипулятивные формулировки;
* не ставить медицинские диагнозы;
* не упоминать данные здоровья или календаря слишком явно, если это может выглядеть пугающе или некомфортно;
* всегда давать пользователю возможность отключить проактивные сообщения;
* если пользователь находится в кризисном состоянии, использовать отдельный safety protocol, а не оптимизацию через бандита.

**Comment by NikPeg** [comment]
Подумать что делать с неактивными пользователями - интуитивно надо писать реже
